### DISCUSSION TITLE
Automating playlist import via API: Possible, documented somewhere?

### DISCUSSION THREAD

**Discussion by Moonbase59** [AzuraCast/AzuraCast#6487] [Q&A] [answered]
I run a test & evaluation station locally, with ~157k songs (from a central music store, NFS mount), and ~60 playlists. I’ve checked out ways for more playlist update automation, since, to be honest, I’m sick and tired of manually updating playlists in Clementine, Strawberry or Quod Libet every few days, exporting them using relative paths, logging into AzuraCast and manually creating copies of >50 "show block" playlists (to get them empty), then importing, deleting the old playlist, and renaming the copy to the old name…

So I played around a little and now at least have automated playlist updating and generation on my end. I think it must be possible to use AzuraCasts great API to drive that even further—use predefined search queries on my database and somehow bring AzuraCast to empty & re-import defined playlists via the API, possibly using Python.

In Swagger, I found an API endpoint
```
https://example.com/api/station/1/playlist/149/import
```
but no further documentation.

**Giving above workflow for automated playlist updates, how would I use the AzuraCast API to facilitate that? What must be gotten, what must be POSTed, and how is the control flow and data format? Any docs or examples available? Or at least some hints where I should have a loook in the code?**

I currently have `.m3u`-like newline-separated relative file path data with the same starting (mount) point that the station’s  `media` folder sees, and the filenames are exact and UTF-8-encoded (but _not_ URL-encoded). These can already be manually imported perfectly.

I also have an API key with correct rights (which I use for my video stream API interfacing) but only use GETs so far and handle the rest in Liquidsoap.

**Does the API need any other format or can `import` handle simple file path lists?**

Help much appreciated. Code will be FOSS, of course—whenever I’ve successfully hacked it together…

For a first step, I plan to use "saved search queries" from Quod Libet, and a Python program to handle the interfacing. All Linux. This will (hopefully) allow just updating the Quod Libet database whenever new music comes in, and have my Python run complex queries for "intelligent" playlists which shall then automagically arrive in my station playlists. Which, in turn, will save _much_ work and time.

**Comment by BusterNeece** [comment, marked as answer]
@Moonbase59 You can just send a M3U playlist to that endpoint via a multipart `POST` form request with the file name parameter being `playlist_file`.

Indeed, all it has to be is just a list of filenames separated by a newline and we can parse that.

How we do this in Vue is pretty simple:

```html
<input type="file" @change="uploaded" />
```

```javascript
const uploaded = (file) => {
  const formData = new FormData();
  formData.append('playlist_file', file);
  axios.post(importPlaylistUrl.value, formData).then((resp) => {
    if (resp.data.success) {
      // Handle Success
    } else {
      // Handle Failure
    }
  });
};
```

**Comment by Moonbase59** [comment]
Thanks, that’s helpful! I’ll certainly experiment with that…
